db/dcrpg/internal/txstmts.go

```diff
@@ -55,36 +57,52 @@ const (
 		num_vin INT4,
 		vin_db_ids INT8[],
 		num_vout INT4,
-		vout_db_ids INT8[]
+		vout_db_ids INT8[],
+		is_valid BOOLEAN,
+		is_mainchain BOOLEAN
 	);`
 
-	SelectTxByHash       = `SELECT id, block_hash, block_index, tree FROM transactions WHERE tx_hash = $1;`
-	SelectTxsByBlockHash = `SELECT id, tx_hash, block_index, tree, block_time FROM transactions WHERE block_hash = $1;`
+	SelectTxByHash = `SELECT id, block_hash, block_index, tree
+		FROM transactions WHERE tx_hash = $1;`
+	SelectTxsByBlockHash = `SELECT id, tx_hash, block_index, tree, block_time
+		FROM transactions WHERE block_hash = $1;`
 
 	SelectTxBlockTimeByHash = `SELECT block_time FROM transactions where tx_hash = $1 
 		ORDER BY block_time DESC LIMIT 1;`
 
 	SelectTxIDHeightByHash = `SELECT id, block_height FROM transactions WHERE tx_hash = $1;`
 
-	SelectTxsPerDay = `SELECT to_timestamp(time)::date as date, count(*) FROM transactions GROUP BY date ORDER BY date;`
+	SelectTxsPerDay = `SELECT to_timestamp(time)::date as date, count(*) FROM transactions
+		GROUP BY date ORDER BY date;`
 
 	SelectFullTxByHash = `SELECT id, block_hash, block_height, block_time, 
 		time, tx_type, version, tree, tx_hash, block_index, lock_time, expiry, 
-		size, spent, sent, fees, num_vin, vin_db_ids, num_vout, vout_db_ids 
+		size, spent, sent, fees, num_vin, vin_db_ids, num_vout, vout_db_ids,
+		is_valid, is_mainchain
 		FROM transactions WHERE tx_hash = $1;`
 
-	SelectTicketsOutputCountByAllBlocks = `SELECT block_height,
-		SUM(CASE WHEN num_vout = 3 THEN 1 ELSE 0 END) as solo,
-		SUM(CASE WHEN num_vout = 5 THEN 1 ELSE 0 END) as pooled
-		FROM transactions WHERE tx_type = 1 GROUP BY block_height
-		ORDER BY block_height;`
+	SelectTxnsVinsByBlock = `SELECT vin_db_ids, is_valid, is_mainchain
+		FROM transactions WHERE block_hash = $1;`
 
-	SelectTicketsOutputCountByTPWindow = `SELECT
-		floor(block_height/144) as count,
-		SUM(CASE WHEN num_vout = 3 THEN 1 ELSE 0 END) as solo,
-		SUM(CASE WHEN num_vout = 5 THEN 1 ELSE 0 END) as pooled
-		FROM transactions WHERE tx_type = 1
-		GROUP BY count ORDER BY count;`
+	UpdateRegularTxnsValidMainchainByBlock = `UPDATE transactions
+		SET is_valid=$1, is_mainchain=$2 
+		WHERE block_hash=$3 and tree=0;`
+
+	UpdateRegularTxnsValidByBlock = `UPDATE transactions
+		SET is_valid=$1 
+		WHERE block_hash=$2 and tree=0;`
+
+	UpdateTxnsMainchainByBlock = `UPDATE transactions
+		SET is_mainchain=$1 
+		WHERE block_hash=$2;`
+
+	UpdateTxnsValidMainchainAll = `UPDATE transactions
+		SET is_valid=b.is_valid, is_mainchain=b.is_mainchain
```

Comment:
Because of the subquery into blocks?

But this will be use in two case: reorg that orphans a block, and block invalidated by stakeholders (votes in following block).  Neither cases where we can tolerate it being very slow, but not called by any user-accessible API.

Reply:
Because of the subquery into blocks?

But this will be use in two case: reorg that orphans a block, and block invalidated by stakeholders (votes in following block). Neither cases where we can tolerate it being very slow, but not called by any user-accessible API.

Reply:
ok. Noted